DOC Clean up the PR checklist (#34239)
scikit-learn/scikit-learn: doc/developers/contributing.rst

Author: If I'm not mistaken, CI complains if `codecov` decreases from main, so it actually requires 100% for new code.

Reviewer: And in another place in this document, we write we require 90% coverage.
What do you think about removing this entirely from here? People will see it when they check the CI and see not all tests have passed.

Author: This would also go into the "contributing a new feature" part.

Author: This is already part of the PR checklist, so I would remove it here. The second point  below could go into the coding guidelines, but the other two I would actually add to the checklist as separate points.

Reviewer: Now that we recommend `pre-commit`, having information on `mypy` here may not be very important anymore. I would remove it as well.

Author: I removed the parts on `coverage` and `mypy` here, but I'm not sure if this kind of information should be kept somewhere for (new) maintainers? I would leave this for another follow-up, though.

Reviewer 2: I think this a typo, and should be "starring it"?

Reviewer 2: Also, could I suggest removing the semicolon. Suggested fix:

"I use it"_, which_ helps us promote...

Reviewer 2: move to new PR

Reviewer 2: more natural to say "that others _have_ reported"

Reviewer 2: moved to new PR

Reviewer 2: comma not needed

Reviewer 2: moved to new PR

Reviewer 3: I think this part could be rephrased. "submitting a ticket to the bug tracker" sounds weird to me, I would find it clearer to simply say "do not hesitate to open an issue to describe your problem". Also I am not fond of "you are welcome to post pull requests", as I believe the desired workflow is to open an issue and wait for a decision to be reached before opening a PR. Also one does not _post_ a PR but _opens_ a PR :p

Author: I haven't looked at that part so far (there are still a few things that could be improved in our contributing guidelines), but I agree that this sounds weird and basically only repeats the information from the paragraph before. I'm not sure if I want to touch this within this PR, but for a follow-up (would you maybe like to do this, @FrancoisPgm?):
How about removing this paragraph and simply linking to the issue tracker in line 181 above (and maybe also adding a link to the FAQ on the feature inclusion requirements in line 182)?

Reviewer 3: > I haven't looked at that part so far

Yes I noticed that wasn't part of your diff, but since I was reading the whole page rendered I figured I could give feedback here. But it's true that this PR is focused on the checklist, I can open a follow-up PR for this part. 

> How about removing this paragraph and simply linking to the issue tracker in line 181 above (and maybe also adding a link to the FAQ on the feature inclusion requirements in line 182)?

I like this suggestion, indeed the paragraph doesn't bring any new information compared to the previous one.

Reviewer 4: Feel free to open a separate issue to track this one 🙏!

#. :ref:`setup_development_environment`
#. Find an issue to work on (see :ref:`new_contributors`)
#. Follow the :ref:`development_workflow`
#. Make sure, you noted the :ref:`pr_checklist`

> If you want to contribute :ref:`contribute_documentation`,

Author: This is now part of the PR checklist for Documentation.

  already said they'd be working on the issue in the past 2-3 weeks, please let
  them finish their work, otherwise consider it stalled and take it over.

To maintain the quality of the codebase and ease the review process, any
contribution must conform to the project's :ref:`coding guidelines

> <coding-guidelines>`, in particular:

Author: Point 1 and 3 below are now part of the PR checklist, point 2 is part of the coding guidelines (which are referenced in the checklist)

  the "why" rather than the "what".
- **Most importantly**: Do not contribute code that you don't understand.

.. seealso::

   For two very well documented and more detailed guides on development
   workflow, please pay a visit to the `Scipy Development Workflow
   <https://scipy.github.io/devdocs/dev/dev_quickstart.html>`_ -
   and the `Astropy Workflow for Developers
   <https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html>`_
   sections.

> 

Reviewer: I think I would leave this away. Why would contributors care about who other projects are set up? Also, in the meantime, our doc are better. :)
```suggestion
```

Pull request checklist
----------------------

.. div:: contribution-types

  .. tab-set::
    :class: contrib-type

    .. tab-item:: Code
      :class-label: tab-6

      .. raw:: html

>         <h4 id="pr-checklist-code-before">Before opening a PR, make sure that:</h4>

Reviewer 5: Please see my comment about headers.
thanks

        * New features have some maintenance overhead. We expect PR authors
          to take part in the maintenance for the code they submit, at least
          initially.

        * New features need to be illustrated with narrative
          documentation in the user guide, with small code snippets.
          If relevant, please also add references in the literature, with PDF links
          when possible. The user guide should also include expected time and space complexity
          of the algorithm and scalability, e.g. "this algorithm can scale to a
          large number of samples > 100000, but does not scale in dimensionality:
          `n_features` is expected to be lower than 100".

        * New features should also include examples of usage. Have a look at other
          examples in the `examples
          <https://github.com/scikit-learn/scikit-learn/tree/main/examples>`__ directory
          for reference. Examples should demonstrate why the new functionality is useful
          in practice and, if possible, compare it to other methods available in
          scikit-learn.

      .. raw:: html

        <h4 id="pr-checklist-code-when">When opening a PR</h4>

> 

Reviewer 5: My comment is getting lost. It may show twice, but I need to say this:

Thanks for removing the uppercase.
Now, for the headings, Sphinx normally takes care of them, and the standard way (in scikit-learn) is just to use =, _, ~. And it also gets the structure (hierarchy) from these.

I found this for general info: https://lpn-doc-sphinx-primer-devel.readthedocs.io/concepts/heading.html
but I think in scikit-learn only uses the symbols I mention above. I suggest you see how it's done on main.

I would avoid adding more to the .. raw:: html section.
I couldn't comment earlier (before you committed). It's a holiday today.

Author: I tried that, but this doesn't work inside a `tab-item`. It gives a `CRITICAL: Unexpected section title.`
and the content is not rendered at all. 
According to Claude "real RST section headers can't legally live inside tab-item (or most directives) — docutils' `nested_parse` disallows titles there by default, which is why your attempt threw `CRITICAL: Unexpected section title.` Raw HTML is the only way to get a genuine heading tag in that spot."
So if we want a real headline (anchored with a unique id) instead of just a boldfaced beginning of a paragraph, I think we need to use the html here. Or do you know any other way we could try?

(This applies to all other places in the file where this is used as well.)

Reviewer 5: Maybe we'll have to use the bold type then. I would really like to avoid adding to custom.css unless absolutely necessary. WDYT?

We already have a complicated system: the css style sometimes is set by Sphinx, or by the pydata theme, by pandas, etc etc. 

BTW: all the sections like these fail for accessibility:

<img width="803" height="82" alt="Screenshot 2026-07-14 at 17 51 32" src="https://github.com/user-attachments/assets/d805fd8e-f661-4a59-b9d8-f704d282607c" />

Reviewer 5: And why do you need "ids" for the headers?

Author: I added those because you mentioned 
>People can use anchor links

above. From what I understand, his requires a unique id.

Author: >BTW: all the sections like these fail for accessibility

Can you be a bit more specific? Is there something we can adapt to make such sections accessible or is this kind of hidden content itself the problem?

Reviewer 5: It's not about contrast. The accessibility report shows this:

<img width="622" height="141" alt="Screenshot 2026-07-15 at 11 54 16" src="https://github.com/user-attachments/assets/f094998d-1f17-4ff5-a58f-3af87095d64b" />

It looks like this is set by Sphinx itself, so that would be an upstream issue. 

I don't know the exact fix until I try it (the "fix" would be added to the `custom.css`). This needs investigation. 

Or we could remove what's under them and show them like normal text. Add a "TODO" as a future improvement and report it to Sphinx.

If you/people decide to keep them, they would add to the long list of things to fix on this topic. Which is not so good.

Reviewer 5: I wasn't specific earlier because I wasn't done.

Reviewer 5: >Me: And why do you need "ids" for the headers?
>You: I added those because you mentioned "People can use anchor links"
>You: above. From what I understand, his requires a unique id.

That `id` was incorrect CSS. We either need a class or an id, but not both for the same purpose. Plus the `id` was not being used anywhere. Sometimes people use LLMs to write CSS.  But the code generated needs a "human in the loop". It may look "OK" but it's either bloated, fragile, adds dead code, or it's difficult to maintain.

Author: You are right in that I am not an expert in css, I'm here to learn. But as the "human in the loop", I did check that this did not raise any errors/warnings when building it locally and that adding the `id` was needed to link to the specific heading (which was not used anywhere specifically, but as I mentioned above, that was my understanding of what was required for making it possible for people to link to it). And at least according to [these guidelines](https://developers.google.com/style/accessibility) (in lack of a better reference), this is acceptable, not "incorrrect", but I do know now that there are other recommended options, in case this comes up again in the future.

Reviewer 5: Thank you for taking my comments into account and having removed the CSS.

Reviewer 5: >You are right in that I am not an expert in css

I never said that. My point was different.

Reviewer 5: >But as the "human in the loop", I did check that this did not raise any errors/warnings when building it locally and that adding the id was needed to link to the specific heading (which was not used anywhere specifically, but as I mentioned above, that was my understanding of what was required for making it possible for people to link to it). And at least according to [these guidelines](https://developers.google.com/style/accessibility) (in lack of a better reference), this is acceptable, not "incorrrect", but I do know now that there are other recommended options, in case this comes up again in the future.

Just to be clear, I didn't comment about this earlier because the CSS part that was incorrect, has been removed from the code. The whole CSS is gone, I mean. Plus what you mention is different of the topic I was talking about. 

Last but not least, I don't agree with that statement. But it's absolute irrelevant (if I agree or not).

        one is preceded by a keyword). **Upon merging, those issues/PRs will
        automatically be closed by GitHub**. If your pull request is simply
        related to some other issues/PRs, or it only partially resolves the target
        issue, create a link to them without using the keywords (e.g., ``Towards #1234``).

      * An incomplete contribution -- where you expect to do more work before receiving
        a full review -- should be marked as a `draft pull request
        <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request>`__
        and changed to "ready for review" when it matures. Draft PRs may be useful to:
        indicate you are working on something to avoid duplicated work, request
        broad review of functionality or API, or seek collaborators. Draft PRs often
        benefit from the inclusion of a `task list
        <https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments>`_ in
        the PR description.

      .. raw:: html

        <h4 id="pr-checklist-code-after">After opening a PR</h4>

> 

Reviewer 5: Same here. I would use the standard Sphinx/scikit-learn way (using `=` or `-` etc, depending on the level you need).

      * Add a `changelog entry
        <https://github.com/scikit-learn/scikit-learn/blob/main/doc/whats_new/upcoming_changes/README.md>`_
        describing the changes if the PR is likely to affect users.

      * Make sure all CI tests pass (check the provided links for more information on
        any errors).

      * If any part of the documentation was changed (which is very likely), also
        check the rendered docs, see :ref:`generated_doc_CI`.

      * Be patient: Before a PR can be merged, it needs to be approved by **two** core
        developers.

    .. tab-item:: Documentation
      :class-label: tab-6

      .. raw:: html

        <h4 id="pr-checklist-docs-before">Before opening a PR</h4>

> 

Reviewer 5: Same here

      * Unless you're only fixing typos or obvious inconsistencies in the
        documentation, make sure the PR relates to at least one existing issue
        that is ready for contribution. Please **do not** open PRs for issues:

        - that are labeled with :ref:`"Needs triage" <issues_tagged_needs_triage>` or any
          other kind of "Needs..." labels,
        - where the discussion has not settled down to an explicit resolution plan,
        - where the reporter has already expressed interest in opening a PR, or
        - that already have cross-referenced and active PRs.

      * Make sure you followed the guidelines on contributing
        :ref:`contribute_documentation`.

      * Check that you can :ref:`build the documentation locally <building_documentation>`.

      .. raw:: html

        <h4 id="pr-checklist-docs-when">When opening a PR</h4>

> 

Reviewer 5: Same here

        one is preceded by a keyword). **Upon merging, those issues/PRs will
        automatically be closed by GitHub**. If the pull request is simply
        related to some other issues/PRs, or it only partially resolves the target
        issue, create a link to them without using the keywords (e.g., ``Towards #1234``).

      * An incomplete contribution -- where you expect to do more work before receiving
        a full review -- should be marked as a `draft pull request
        <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request>`__
        and changed to "ready for review" when it matures. Draft PRs may be useful to:
        indicate you are working on something to avoid duplicated work, request
        broad review of functionality or API, or seek collaborators. Draft PRs often
        benefit from the inclusion of a `task list
        <https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments>`_ in
        the PR description.

      .. raw:: html

        <h4 id="pr-checklist-docs-after">After opening a PR</h4>

> 

Reviewer 5: And finally, same comment about the headers.

Pull request checklist
----------------------

> Before a PR can be merged, it needs to be approved by two core developers.

Author: This is now at the bottom of the checklist.

      }

      .contribution-types .sd-tab-set.contrib-type::before {
        content: "Type of Contribution";
      }

    }
    p.rubric.rubric-large {
      font-size: 1.3em;
      border-bottom: none;
    }
    .contribution-types li {
      list-style-type: "✓ ";
    }
    .contribution-types li li {
      list-style-type: "✗ ";
    }
  </style>

.. div:: contribution-types

  .. tab-set::
    :class: contrib-type

    .. tab-item:: Code
      :class-label: tab-6

      .. rubric:: BEFORE opening a PR, make sure that:
         :class: rubric-large

      * Your PR relates to at least one existing issue that is ready for
        contribution.
        Please **do not** open PRs for issues:

        - that are labeled with :ref:`"Needs triage" <issues_tagged_needs_triage>` or any
          other kind of "Needs..." labels,
        - where the discussion has not settled down to an explicit resolution plan,
        - where the reporter has already expressed interest in opening a PR, or
        - that already have cross-referenced and active PRs.

        If no related issue exists yet, please :ref:`open one <submitting_bug_feature>`
        to discuss your proposal first.

>       * You understand the code that you are about to submit.

Reviewer 5: On the original document, this point was very prominent. It now loses the emphasis that signaled it as the single most critical rule. 
<img width="688" height="242" alt="Image" src="https://github.com/user-attachments/assets/ff087386-48bb-46a6-9505-ef7bc700b9b1" />

Author: I integrated the highlighting into the checklist.

        <https://github.com/blog/1506-closing-issues-via-pull-requests/>`_
        (e.g., ``Fixes #1234``; multiple issues/PRs are allowed as long as each
        one is preceded by a keyword). **Upon merging, those issues/PRs will
        automatically be closed by GitHub**. If your pull request is simply
        related to some other issues/PRs, or it only partially resolves the target
        issue, create a link to them without using the keywords (e.g., ``Towards #1234``).

      .. rubric:: AFTER opening a PR
         :class: rubric-large

      * Add a `changelog entry
        <https://github.com/scikit-learn/scikit-learn/blob/main/doc/whats_new/upcoming_changes/README.md>`_
        describing the changes if the PR is likely to affect users.

      * Make sure all CI tests pass (check the provided links for more information on
        any errors).

      * If any part of the documentation was changed (which is very likely), also
        check the rendered docs, see :ref:`generated_doc_CI`.

    .. tab-item:: Documentation
      :class-label: tab-6

      .. rubric:: BEFORE opening a PR

>          :class: rubric-large

Reviewer 5: I would avoid using uppercase -- some people may take it as aggressive (like shouting). The intention here  I imagine, is a replacement of headings. But true headings have benefits that uppercase don't: 

- People can use anchor links
- Screen-readers can jump between sections
- Consistent visual hierarchy with the rest of the page

Note: if you use different headings than what you have, they should be hierarchical (for accessibility reasons and standard UX rules).

Author: I removed the uppercasing. 
From what I understand, we could convert them to proper headings by replacing the `..rubric::` part with this:
```html
        .. raw:: html
          <h4 id="pr-checklist-code-before">Before opening a PR, make sure that:</h4>
```
Then we will also not need the `rubric-large` definition above. 

It then looks like this:
<img width="831" height="627" alt="image" src="https://github.com/user-attachments/assets/103f996f-ee16-4096-a934-17f5b8691307" />

It could be nice to try and add some space above the first headline, but that would again require inflating the css config, so I'm not sure we want that.

Reviewer 5: Thanks for removing the uppercase. 
Now, for the headings, Sphinx normally takes care of them, and the standard way (in scikit-learn) is just to use `=`, `_`, `~`.  And it also gets the structure (hierarchy) from these.

I found this for general info: https://lpn-doc-sphinx-primer-devel.readthedocs.io/concepts/heading.html
but I think in scikit-learn only uses the symbols I mention above. I suggest you see how it's done on main.

I would avoid adding more to the `.. raw:: html` section. 
I couldn't comment earlier (before you committed). It's a holiday today.

Author: I added the changes I proposed here and moved all the styling to `custom.css`, so the source file is clean, but still looks nice in the rendered version. :)

Reviewer 5: There are two things: one is adding the styling to custom.css, and the other is the type of headers that you are adding. For the headers, please see my comment above. https://github.com/scikit-learn/scikit-learn/pull/34239#discussion_r3578430426

        related to some other issues/PRs, or it only partially resolves the target
        issue, create a link to them without using the keywords (e.g., ``Towards #1234``).

      .. rubric:: AFTER opening a PR
         :class: rubric-large

      * Make sure all CI tests pass (check the provided links for more information on
        any errors).

      * Check the rendered docs, see :ref:`generated_doc_CI`.

Before a PR can be merged, it needs to be approved by **two** core developers.
An incomplete contribution -- where you expect to do more work before receiving
a full review -- should be marked as a `draft pull request
<https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request>`__
and changed to "ready for review" when it matures. Draft PRs may be useful to:
indicate you are working on something to avoid duplicated work, request
broad review of functionality or API, or seek collaborators. Draft PRs often
benefit from the inclusion of a `task list
<https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments>`_ in

> the PR description.

Reviewer 5: Is this section meant to be an appendix to the checklist? (the section "Before a PR can be merged...")
If this was intended, that's fine but it reads a bit odd:

<img width="766" height="328" alt="Image" src="https://github.com/user-attachments/assets/428637a5-e537-4e86-be1c-2cd8bf83ebcd" />

Author: It actually used to come before the checklist. I thought it makes more sense to add it afterwards, because it did read like an appendix to me, but I'm also fine with putting it at the beginning again.
WDYT?

Author: Alternatively, we could also integrate the part on draft PRs into the `When opening a PR` section and the note on the two reviewers into the `After opening a PR`?

Author: I went with my last suggestion above and integrated the two points into the checklist.

Reviewer 5: I vote for the second option.

Pull request checklist
----------------------

.. div:: contribution-types

  .. tab-set::
    :class: contrib-type

    .. tab-item:: Code
      :class-label: tab-6

      **Before** opening a PR, make sure that:

      * Your PR relates to at least one existing issue that is ready for
        contribution.
        Please **do not** open PRs for issues:

        - that are labeled with :ref:`"Needs triage" <issues_tagged_needs_triage>` or any
          other kind of "Needs..." labels,
        - where the discussion has not settled down to an explicit resolution plan,
        - where the reporter has already expressed interest in opening a PR, or
        - that already have cross-referenced and active PRs.

        If no related issue exists yet, please :ref:`open one <submitting_bug_feature>`
        to discuss your proposal first.

>       * **Most importantly**: You understand the code that you are about to submit.

Reviewer 4: ```suggestion
      * **Most importantly**: you need to understand the code that you are about to submit.
```

        developers.

    .. tab-item:: Documentation
      :class-label: tab-6

      **Before** opening a PR:

      * Unless you're only fixing typos or obvious inconsistencies in the
        documentation, make sure the PR relates to at least one existing issue
        that is ready for contribution. Please **do not** open PRs for issues:

        - that are labeled with :ref:`"Needs triage" <issues_tagged_needs_triage>` or any
          other kind of "Needs..." labels,
        - where the discussion has not settled down to an explicit resolution plan,
        - where the reporter has already expressed interest in opening a PR, or
        - that already have cross-referenced and active PRs.

      * Make sure you followed the guidelines on contributing
        :ref:`contribute_documentation`.

      * Check that you can :ref:`build the documentation locally <building_documentation>`.

      **When** opening a PR:

      * Give the pull request a helpful title, starting with "DOC", that summarizes
        what the contribution does. This title will become the commit message once
        merged.

>       * Fill out the `pull request template <https://github.com/scikit-learn/scikit-learn/blob/main/.github/PULL_REQUEST_TEMPLATE.md>`_.

Reviewer 4: I am wondering whether it is slightly better to link to the raw markdown (rather than rendered) which contains instructions?

```suggestion
      * Fill out the `pull request template <
https://raw.githubusercontent.com/scikit-learn/scikit-learn/refs/heads/main/.github/PULL_REQUEST_TEMPLATE.md>`_.
```

Author: I agree, but I think you accidentally added a line break that broke the doc build.
```
'/home/circleci/project/doc/developers/contributing.rst:510: WARNING: Inline interpreted text or phrase reference start-string without end-string. [docutils]
/home/circleci/project/doc/developers/contributing.rst:511: WARNING: Explicit markup ends without a blank line; unexpected unindent. [docutils]'
```

Author: I added a fix, but this disabled auto-merge. Could you please enable it again?

      }

    }
    p.rubric.rubric-large {
      font-size: 1.3em;
      border-bottom: none;
    }
    .contribution-types li {
      list-style-type: "✓ ";
    }
    .contribution-types li li {
      list-style-type: "✗ ";
    }
  </style>

.. div:: contribution-types

  .. tab-set::
    :class: contrib-type

    .. tab-item:: Code
      :class-label: tab-6

      .. rubric:: BEFORE opening a PR, make sure that:
         :class: rubric-large

      * The PR relates to at least one existing issue that is ready for
        contribution.
        Please **do not** open PRs for issues:

        - that are labeled with `"Needs triage" <issues_tagged_needs_triage>`_ or any
          other kind of "Needs..." labels,
        - where the discussion has not settled down to an explicit resolution plan,
        - where the reporter has already expressed interest in opening a PR, or
        - that already have cross-referenced and active PRs.

        If no related issue exists yet, please :ref:`open one <submitting_bug_feature>`
        to discuss your proposal first.

      * You followed the :ref:`coding-guidelines` and properly commented and
        documented your code.

>       * When applicable, you used the validation tools and scripts in the

Author: I would move this part into the coding guidelines to keep this list more concise.

      Before a PR can be merged, it needs to be approved by **two** core developers.
      An incomplete contribution -- where you expect to do more work before receiving
      a full review -- should be marked as a `draft pull request
      <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request>`__
      and changed to "ready for review" when it matures. Draft PRs may be useful to:
      indicate you are working on something to avoid duplicated work, request
      broad review of functionality or API, or seek collaborators. Draft PRs often
      benefit from the inclusion of a `task list
      <https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments>`_ in
      the PR description.

    .. tab-item:: Documentation
      :class-label: tab-6

      .. rubric:: BEFORE opening a PR
         :class: rubric-large

      * If the changes are mainly fixing typos or obvious inconsistencies in the
        documentation, ignore this point.
        Otherwise, make sure the PR relates to at least one existing issue that is

>         ready for contribution. Please **do not** open PRs for issues:

Reviewer: ```suggestion
      * Unless you're only fixing typos or obvious inconsistencies in the
        documentation, make sure the PR relates to at least one existing issue 
        that is ready for contribution. Please **do not** open PRs for issues:
```
Bit shorter.

         :class: rubric-large

      * Give the pull request a helpful title, starting with "DOC", that summarizes
        what the contribution does. This title will become the commit message once
        merged.

      * Fill out the `pull request template <https://github.com/scikit-learn/scikit-learn/blob/main/.github/PULL_REQUEST_TEMPLATE.md>`_.

      * If the pull request resolves any issues/PRs, use `keywords to create a link to them
        <https://github.com/blog/1506-closing-issues-via-pull-requests/>`_
        (e.g., ``Fixes #1234``; multiple issues/PRs are allowed as long as each
        one is preceded by a keyword). **Upon merging, those issues/PRs will
        automatically be closed by GitHub**. If the pull request is simply
        related to some other issues/PRs, or it only partially resolves the target
        issue, create a link to them without using the keywords (e.g., ``Towards #1234``).

      .. rubric:: AFTER opening a PR
         :class: rubric-large

      * Make sure all CI tests pass (check the provided links for more information on
        any errors).

      * Check the rendered docs, see :ref:`generated_doc_CI`.

>       Before a PR can be merged, it needs to be approved by **one** core developer.

Reviewer: ```suggestion
      Before a PR can be merged, it needs to be approved by **one or two** core 
      developers.
```
It's mostly two, except in very simple cases.

Reviewer 6: I think I would stick with two, the same as what we have in code. I don't want to give the impression that documentation changes require less review/is easier - it's hard to write good documentation!

      An incomplete contribution -- where you expect to do more work before receiving
      a full review -- should be marked as a `draft pull request
      <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request>`__
      and changed to "ready for review" when it matures. Draft PRs may be useful to:
      indicate you are working on something to avoid duplicated work, request
      broad review of functionality or API, or seek collaborators. Draft PRs often
      benefit from the inclusion of a `task list
      <https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments>`_ in
      the PR description.

> 3. **Make sure your code passes the tests**. The whole test suite can be run

Author: I think I would rather move this section into a separate subsection on this site and link to it from the checklist above instead of adding all these details above.

  on all new contributions will get the overall code base quality in the

>   right direction.

Reviewer: I feel this note rather belongs to the "Coding guidelines" section (which in a future PR is supposed to get its own page/section inside the contribution guidelines).

  on all new contributions will get the overall code base quality in the
  right direction.

> .. seealso::

Author: @StefanieSenger and I agreed that our docs are shaping to be good enough so that we don't need to link to other projects for this anymore :)

   you don't need to run the whole test suite locally. For guidelines on how
   to use ``pytest`` efficiently, see the :ref:`pytest_tips`.

> 9. PRs should often substantiate the change, through benchmarks of

Author: I think this could go into a (drop-down?) `Contributing a new feature` section in the `Code - BEFORE PR` part.

Reviewer: And including a link to "What are the inclusion criteria for new algorithms" in `faq.rst`.

Reviewer 6: You probably know this already but these items e.g. 9 onwards appear under the 'Documentation' tab but I am not sure they are relevant?

Author: They are now integrated into the `Code` tab.

High-quality `unit testing <https://en.wikipedia.org/wiki/Unit_testing>`_
is a corner-stone of the scikit-learn development process. For this
purpose, we use the `pytest <https://docs.pytest.org>`_
package.

The tests are functions appropriately named, located in `tests`
subdirectories, that check the validity of the algorithms and the
different options of the code.

The whole test suite can be run with `pytest`, but it is usually not recommended since

> it takes a long time. It is often enough to only run the test related to your changes:

Reviewer 4: ```suggestion
The whole test suite can be run with `pytest` and takes some time, roughly 10-20 minutes. It is often enough to only run the test related to your changes:
```

It's feel a bit weird to say "we don't recommend to run the full test suite, so I tweaked the wording a bit.

High-quality `unit testing <https://en.wikipedia.org/wiki/Unit_testing>`_
is a corner-stone of the scikit-learn development process. For this
purpose, we use the `pytest <https://docs.pytest.org>`_
package. The tests are functions appropriately named, located in `tests`
package.

The tests are functions appropriately named, located in `tests`
subdirectories, that check the validity of the algorithms and the
different options of the code.

Running `pytest` in a folder will run all the tests of the corresponding

> subpackages. For a more detailed `pytest` workflow, please refer to the

Author: I integrated the details from the PR checklist into this section and the PR checklist now contains a link to this here instead.

`sklearn/linear_model/_logistic.py`, running the following commands will
usually be enough:

- `pytest sklearn/linear_model/_logistic.py` to make sure the doctest
  examples are correct
- `pytest sklearn/linear_model/tests/test_logistic.py` to run the tests
  specific to the file
- `pytest sklearn/linear_model` to test the whole
  :mod:`~sklearn.linear_model` module
- `pytest doc/modules/linear_model.rst` to make sure the user guide
  examples are correct.
- `pytest sklearn/tests/test_common.py -k LogisticRegression` to run all our
  estimator checks (specifically for `LogisticRegression`, if that's the
  estimator you changed).

There may be other failing tests, but they will be caught by the CI so
you don't need to run the whole test suite locally. For guidelines on how

> to use ``pytest`` efficiently, see the :ref:`pytest_tips`.

Author: As another defragmentation follow-up, this could be combined with the referenced `pytest_tips` section into a separate testing page/section.

Reviewer 7: I wouldn't mind having a separate page for `pytest` materials.
